src/main/java/de/tum/in/www1/artemis/web/rest/push_notification/PushNotificationResource.java

```diff
@@ -0,0 +1,117 @@
+
+import de.tum.in.www1.artemis.config.Constants;
+import de.tum.in.www1.artemis.domain.User;
+import de.tum.in.www1.artemis.domain.push_notification.PushNotificationDeviceConfiguration;
+import de.tum.in.www1.artemis.domain.push_notification.PushNotificationDeviceConfigurationId;
+import de.tum.in.www1.artemis.repository.PushNotificationDeviceConfigurationRepository;
+import de.tum.in.www1.artemis.repository.UserRepository;
+import io.jsonwebtoken.*;
+
+/**
+ * Rest Controller for managing push notification device tokens for native clients.
+ */
+@RestController
+@RequestMapping("/api/push_notification")
+public class PushNotificationResource {
+
+    private static KeyGenerator aesKeyGenerator;
+
+    static {
+        try {
+            aesKeyGenerator = KeyGenerator.getInstance("AES");
+            aesKeyGenerator.init(256, new SecureRandom());
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private PushNotificationDeviceConfigurationRepository pushNotificationDeviceConfigurationRepository;
+
+    private UserRepository userRepository;
+
+    public PushNotificationResource(PushNotificationDeviceConfigurationRepository pushNotificationDeviceConfigurationRepository, UserRepository userRepository) {
+        this.pushNotificationDeviceConfigurationRepository = pushNotificationDeviceConfigurationRepository;
+        this.userRepository = userRepository;
+    }
+
+    /**
+     * API Endpoint which native clients use to register with their device token to enable push notification support
+     *
+     * @param pushNotificationRegisterBody contains all information required to store the device token for a specific user
+     * @return an DTO containing information about the encryption
+     */
+    @PostMapping("register")
+    @PreAuthorize("hasRole('USER')")
+    public ResponseEntity<PushNotificationRegisterDTO> register(@Valid @RequestBody PushNotificationRegisterBody pushNotificationRegisterBody) {
+        var newKey = aesKeyGenerator.generateKey();
+
+        String token = getToken();
+
+        String jwtWithoutSignature = token.substring(0, token.lastIndexOf('.') + 1);
+
+        Jwt<Header, Claims> headerClaimsJwt;
+
+        // This cannot throw an error as it must have been valid to even call this method
+        try {
+            headerClaimsJwt = Jwts.parserBuilder().build().parseClaimsJwt(jwtWithoutSignature);
+        }
+        catch (ExpiredJwtException e) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
+        }
+
+        Date expirationDate = headerClaimsJwt.getBody().getExpiration();
+
+        User user = userRepository.getUser();
+
+        // DB CALL INSERT OR UPDATE
+        pushNotificationDeviceConfigurationRepository.save(new PushNotificationDeviceConfiguration(pushNotificationRegisterBody.token(), pushNotificationRegisterBody.deviceType(),
+                expirationDate, newKey.getEncoded(), user));
```

Comment:
```suggestion
        var configuration = new PushNotificationDeviceConfiguration(pushNotificationRegisterBody.token(), pushNotificationRegisterBody.deviceType(), expirationDate, newKey.getEncoded(), user);
        pushNotificationDeviceConfigurationRepository.save(configuration);
```
- comment does not add value
- (personal preference) long line can be split up into declaration and action to make the sequence clearer